psi4/driver/driver_nbody.py

```diff
@@ -25,830 +25,1291 @@
+
+"""
+
+__all__ = [
+    "ManyBodyComputer",
+]
+
+import copy
 import itertools
 import math
-from typing import Callable, Union
+from typing import Any, Callable, Dict, List, Literal, Optional, Sequence, Set, Tuple, Union
+from ast import literal_eval
+from enum import Enum
+
+from pydantic import Field, validator
+
+import pprint
+pp = pprint.PrettyPrinter(width=120, compact=True, indent=1)
+import logging
 
 import numpy as np
+from qcelemental.models import DriverEnum, AtomicResult
 
 from psi4 import core
-from psi4.driver import p4util
-from psi4.driver import constants
+from psi4.driver import constants, driver_nbody_multilevel, p4util
 from psi4.driver.p4util.exceptions import *
-from psi4.driver import driver_nbody_helper
+from psi4.driver.task_base import BaseComputer, EnergyGradientHessianWfnReturn
 
-### Math helper functions
+logger = logging.getLogger(__name__)
 
+SubTaskComputers = Union["AtomicComputer", "CompositeComputer", "FiniteDifferenceComputer"]
+FragBasIndex = Tuple[Tuple[int], Tuple[int]]
 
-def nCr(n, r):
-    f = math.factorial
-    return f(n) // f(r) // f(n - r)
 
+class BsseEnum(str, Enum):
+    """Available basis-set superposition error (BSSE) treatments."""
 
-### Begin CBS gufunc data
+    nocp = "nocp"  # plain supramolecular interaction energy
+    cp = "cp"      # counterpoise correction
+    vmfc = "vmfc"  # Valiron-Mayer function counterpoise
 
 
-def _sum_cluster_ptype_data(ptype,
-                            ptype_dict,
-                            compute_list,
-                            fragment_slice_dict,
-                            fragment_size_dict,
-                            ret,
-                            vmfc=False,
-                            n=0):
+def _sum_cluster_ptype_data(
+    ptype: DriverEnum,
+    ptype_dict: Dict,
+    compute_list: Set[FragBasIndex],
+    fragment_slice_dict: Dict[int, Sequence],
+    fragment_size_dict: Dict[int, int],
+    mc_level_lbl: int,
+    vmfc: bool = False,
+    nb: int = 0,
+) -> Union[float, np.ndarray]:
     """
-    Sum gradients/hessians from n-body computations to obtain the BSSE corrected or uncorrected gradient/hessian
+    Sum arrays from n-body computations to obtain the BSSE corrected or uncorrected scalar or array.
 
     Parameters
     ----------
-    ptype : str
-        Either "gradient" or "hessian"
+    ptype
+        Hint to shape of array data to sum.
+    ptype_dict
+        Dictionary containing computed energy, gradient, or Hessian obtained from each subsystem computation
+    compute_list
+        A list of (frag, bas) tuples notating all the required computations.
+    fragment_slice_dict
+        Dictionary containing slices that index the gradient or Hessian matrix for each of the 1-indexed fragments.
+        For He--HOOH--Me cluster, `{1: slice(0, 1, None), 2: slice(1, 5, None), 3: slice(5, 10, None)}`.
+    fragment_size_dict
+        Dictionary containing the number of atoms of each 1-indexed fragment.
```

Comment:
What does "1-indexed" mean?

Reply:
as opposed to 0-indexed. so a trimer has frags 1, 2, 3, just as one would use in extract_subsets.